Feature flag, Log, Metric, alternative user ids added

dhavalocked opened the pull request.

noopurprj committed:
- Added set user alias function
- Added Log and metric creation
- Added feature flag function

dhavalocked committed:
Updated README to include definitions of new functions

noopurprj committed:
committed old main.java

noopurprj committed:
Merge branch 'master' of https://github.com/dhavalocked/educational-experiment-service-client-libs

noopurprj committed:
fixed 3 retires on failure

dhavalocked commented:
@jerith666  There are fields in some API's that can be of type `any`. 

For Example: 

-  when we add a metric, We are passing a variable named [`key`](https://github.com/dhavalocked/educational-experiment-service-client-libs#addmetricsmetrics-imetricunit) .  Here type of the `key` variable that server expects is `String or String[]`. I have added this variable as of type `Object`  in [bean](https://github.com/dhavalocked/educational-experiment-service-client-libs/blob/master/java/src/main/java/org/upgradeplatform/requestbeans/MetricUnit.java) class. 

-  Same thing goes with [Log Request object](https://github.com/dhavalocked/educational-experiment-service-client-libs/blob/master/java/src/main/java/org/upgradeplatform/requestbeans/Log.java) where value can be of type `Integer, String or JSON object`. 

1. One way, I am thinking is that we define them as of type Object in bean class. It will allow users to access object with any data type. However, User might have to typecast or add extra `Instance Of` checks. 

2. The other way is to be more strict in type checking, We can define 2 variables `String key, String[] keys` and change server code accordingly.

The drawback of doing it the later way is we should have information about all the possible types that a variable can take in advance. This is the reason i have chosen the  first way in this PR.  

Let me know what are your thoughts.

jerith666 commented:
So the question is basically how best to represent [the `IMetricUnit.key` field](https://github.com/CarnegieLearningWeb/educational-experiment-types/blob/master/src/Experiment/interfaces.ts#L90) in Java, which in TypeScript has the type `string | string[]`?

There are a few libraries that provide an `Either` type for Java, which is what I'd recommend here.  You then may have to do some extra work to ensure that the resulting wire format is compatible, but that's exactly the sort of thing that a good client library _should_ do IMO: convert between the API's wire format and the most natural equivalent in the target language.

For a trinary value (like your second example of `Integer | String | JSON`) you'll end up with a slightly ugly looking nested either, like `Either<Integer,Either<String,JSON>>`.  But I think that's the best we can do in Java without writing a custom class, which seems like overkill.

The first library I would investigate is [Vavr](https://www.vavr.io/); here is [the JavaDoc for its `Either` type](https://www.javadoc.io/doc/io.vavr/vavr/latest/io/vavr/control/Either.html).

jerith666 commented on java/src/main/java/org/upgradeplatform/client/ExperimentClient.java:
```diff
@@ -269,4 +279,158 @@ public void failed(Throwable throwable) {
+
+		AsyncInvoker invocation = this.apiService.prepareRequest(GET_ALL_FEATURE_FLAGS);
+
+		invocation.get(new PublishingRetryCallback<>(invocation, Entity.json(""), MAX_RETRIES, REQUEST_TYPES_GET,
```
Why are you using a `PublishingRetryCallback` here, since you're doing a `get()` and passing in an empty entity?

dhavalocked replied:
`PublishingRetryCallback` is to retry request in case it failed because of temporary network issue or something.  Yeah, i can avoid passing empty entity by creating another constructor in `PublishingRetryCallback `.

dhavalocked replied:
#e7a5725

dhavalocked replied:
Please look at the latest commit.

jerith666 replied:
looks good, thanks!

jerith666 commented on java/src/main/java/org/upgradeplatform/client/ExperimentClient.java:
```diff
@@ -269,4 +279,158 @@ public void failed(Throwable throwable) {
+		return featureFlags.stream()
+				.filter(t -> t.getKey().equals(key))
+				.findFirst()
+				.map(ExperimentClient::getActiveVariation)
```
This is dodgy.  `getActiveVariation()` is not creating a new `FeatureFlag` object, but rather mutating an existing one in the source `featureFlags` list.  That means subsequent callers might see different results.  You're better off making `getActiveVariation()` a pure function that creates a new object, unconnected to the `featureFlag` object passed in to it.

dhavalocked replied:
Yeah true! Updating.

dhavalocked replied:
Please look at the latest commit.

jerith666 replied:
looks good

jerith666 commented on java/src/main/java/org/upgradeplatform/client/ExperimentClient.java:
```diff
@@ -269,4 +279,158 @@ public void failed(Throwable throwable) {
+		return featureFlag;
+	}
+
+	public void setAltUserIds(final String[] altUserIds, final ResponseCallback<List<InitRequest>> callbacks) {
```
Java convention is to avoid arrays in favor of `List`s when possible, unless there's a performance-related reason to use arrays.  (I don't think there is here.)

dhavalocked replied:
Okay! I will make it a list.

dhavalocked replied:
Please look at the latest commit.

jerith666 replied:
looks good

jerith666 commented on java/src/main/java/org/upgradeplatform/client/ExperimentClient.java:
```diff
@@ -269,4 +279,158 @@ public void failed(Throwable throwable) {
+		}));
+	}
+
+	public void addMetrics(final MetricUnit[] metrics, final ResponseCallback<List<Metric>> callbacks) {
```
same comment about lists vs. arrays.

dhavalocked replied:
Please look at the latest commit.

jerith666 replied:
looks good

jerith666 commented on java/src/main/java/org/upgradeplatform/utils/PublishingRetryCallback.java:
```diff
@@ -14,12 +14,14 @@
 	private final Entity<T> message;
 	private final InvocationCallback<Response> callback;
 	private int retries;
+	private String type; // Request type
```
This should be an `enum`, not a `String`

dhavalocked replied:
Yeah Updating it.

dhavalocked replied:
Please look at the latest commit.

jerith666 replied:
nit: the enum should follow standard Java naming conventions: `RequestType` rather than `REQUEST_TYPES`.

jerith666 commented on java/src/main/java/org/upgradeplatform/requestbeans/MetricUnit.java:
```diff
@@ -0,0 +1,65 @@
+	private Metadata metadata;
+	private String[] allowedData;
+	
+	private MetricUnit[]  metricUnit;
```
I don't understand why we have two `MetricUnit[]` fields like this (`children` and `metricUnit`) ... ?

dhavalocked replied:
@VivekFitkariwala

dhavalocked replied:
`MetricUnit[]` was passed as a request body. including it in `MetricUnit.java` was confusing so i have removed it and made a new class. `MetricUnitBody`.

jerith666 replied:
looks good

jerith666 commented on java/src/main/java/org/upgradeplatform/responsebeans/InitRequest.java:
```diff
@@ -12,10 +12,18 @@
 	private Map<String, List<String>> group;
 	private Map<String, String> workingGroup;
-	
+	private InitRequest originalUser;
```
This recursive inclusion of `InitRequest` inside itself doesn't make sense to me.  What is it for?

dhavalocked replied:
@VivekFitkariwala 
When we set alternative userIds, the response object contains  a field `originalUser`.  I have added it to initRequest class so that we don't have to create new class.

Example response object: 

```
{
        "id": "user1alias1",
        "originalUser": {
            "createdAt": "2020-06-11T07:10:21.583Z",
            "updatedAt": "2020-06-11T07:10:21.583Z",
            "versionNumber": 1,
            "id": "user1",
            "group": {
                "teachers": [
                    "1",
                    "2",
                    "7"
                ],
                "classes": [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5"
                ]
            },
            "workingGroup": null
        },
        "group": null,
        "workingGroup": null,
        "createdAt": "2020-06-12T02:55:42.261Z",
        "updatedAt": "2020-06-12T02:55:42.261Z",
        "versionNumber": 1
    }
```

VivekFitkariwala replied:
Whole document of original user might be redundant here. We can just send this data in response.
```
{
        "id": "user1alias1",
        "originalUser": "user1",
        "group": null,
        "workingGroup": null,
        "createdAt": "2020-06-12T02:55:42.261Z",
        "updatedAt": "2020-06-12T02:55:42.261Z",
        "versionNumber": 1
    }
```

noopurprj committed:
- changes suggested in #49 part1
- getActiveVariation() is now a function that creates a copy and returns object rather than changing original
- Used Lists over Arrays wherever possible
- Used enum for REQUEST_TYPES

noopurprj committed:
added new class for sending list of metrics as a body in add metric request

jerith666 commented on java/src/main/java/org/upgradeplatform/utils/PublishingRetryCallback.java:
```diff
@@ -1,25 +1,36 @@
 
 	private final AsyncInvoker invoker;
-	private final Entity<T> message;
+	private Entity<T> message;
```
you can leave this `final` -- you just need to move the `this.message = null;` line into the other constructor :)

noopurprj committed:
#49 Suggested improvements PART 3

noopurprj committed:
Added Either type for Log API

dhavalocked commented:
@jerith666  I have used `Either` type for passing data in Log API. I had to add some extra checks/Conversions to make sure it works. Please have a look. 

I had to add `instance of` checks and assign data accordingly to `Either<Integer, Either<String, JSON>>` as client was not handling it. Have also updated `main` method to show one use case.

jerith666 commented:
So the root of the issue seems to be that you want to pass an arbitrary POJO type (here, `Value`) into the `log()` method, and then expect that the `LogEventResponse` you read back off the wire will embed that exact same kind of POJO (i.e. it will be a `LogEventResponse<Value>`).

But of course there's no guarantee that'll be the case, who knows what might come off the wire?  :)  That's why the compiler is warning you and you have to put those `@SuppressWarnings("unchecked")` annotations in.

This can be made to work with some more effort, but I would step back and ask: why do we care?  Why does the server have to send back to the client _the exact same value that the client just gave it_?  This seems like needless network traffic, and data that the client would never actually care about or need.

So, if you just drop it, I think the problem goes away.

jerith666 commented:
I guess I would also say that I'm not sure this `Either` approach buys us much in the case of `Log.value`.  In the case of `Metric.key`, the possible types are not directly related to each other -- none of them is a super-type of the others.  So it makes sense there.

But in the case of `Log.value`, the three possible types are:

1. `Integer`
1. `String`
1. any arbitrary `Object` than can be converted to JSON

That third type subsumes the other two.  (For example, a `Boolean` would be a valid value, even though it's not one of the primitives we list explicitly.)  So that means there isn't really much use in getting more specific -- it's sufficient to just say "any arbitrary `Object` that can be converted to JSON".

Ideally we'd have an `interface Jsonable` that we could use instead of `Object`, but since we can't retroactively make `java.lang.String` etc. implement that interface, that doesn't work.  (Since there are plenty of subclasses of `Object`, e.g. `Function`, that can't be converted to JSON.)

The closest thing we have is something like [GSON's `JsonElement`](https://javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/JsonElement.html).  It's okay with me if you want to go that route.  But I think it's also okay to just use `Object` in this case, frankly.

noopurprj committed:
Added Object type for logging

noopurprj committed:
Merge remote-tracking branch 'upstream/master'

noopurprj committed:
- Updated types for Add metric request & add log request
- Renamed InitRequest POJO class to ExperimentUser

jerith666 commented on package-lock.json:
```diff
@@ -3612,7 +3612,7 @@
     },
     "upgrade_types": {
-      "version": "github:CarnegieLearningWeb/educational-experiment-types#7726de2644d4006ea4b5ddfb7edc17cbc138674b",
+      "version": "github:CarnegieLearningWeb/educational-experiment-types#67359be6e740d73575ce752810a335966eb6198e",
```
This file isn't Java-related, not sure why it's being touched in this PR ... ?

dhavalocked replied:
Yeah! i pulled this to see some of the interfaces in JS Client lib to make sure they are up to date with JAVA client lib. it was pushed by mistake in this commit. It should not be here.

jerith666 commented on java/src/main/java/org/upgradeplatform/client/ExperimentClient.java:
```diff
@@ -415,23 +432,27 @@ public void failed(Throwable throwable) {
-						e.printStackTrace();
+						callbacks.onError(new ErrorResponse(e.toString()));
 					}
+					callbacks.onSuccess(logEventResponse);
```
This will be invoked even if there's an exception; it should remain up above inside the `try` block.

dhavalocked replied:
Okay!

jerith666 committed:
cleanup generic typing around MetricUnitBody

the <T> type param was unused and there were no raw types here any more

jerith666 committed:
fix control flow in log() to avoid invoking onSuccess() on failure

jerith666 committed:
java lib version bump: 0.1 -> 0.2

jerith666 commented:
See https://github.com/CarnegieLearningWeb/educational-experiment-service-client-libs/tree/metric-log-cleanup for commits that address some of the review issues etc.

VivekFitkariwala commented:
Is this good for merge?

jerith666 commented:
Unless that change to `package-lock.json` will cause problems I don't know about, I think it's good!
